### DISCUSSION TITLE
Companion iOS/Multiplatform app for remote control

### DISCUSSION THREAD

**Discussion by Robertsmania** [vincentneo/LosslessSwitcher#72] [Ideas]
I'd like a companion app to use to manually switch the sample rate and bit depth remotely, either from an iOS device or another Mac computer.

In my setup I have a Mac Mini that runs as a headless media server.  I generally use the iTunes Remote app to control what its playing with my iPad.  

I very much like the LosslessSwitcher features but do find it annoying to hear the hiccup when the rate switches as new songs start, particularly when listening to playlists that have media with different settings. 

My idea is to extend LosslessSwitcher to support a companion app running on another device to interact with the settings.  I'd like to have a manual toggle so the process on the computer running Apple Music would still monitor what is playing and show what it detects for the current media (but not necessarily switch automatically).

That way I could listen in my studio and leave it on a static setting while playing casual playlists or stuff that would often be switching rates.  But if I wanted to listen to something more seriously and set the sample rate and bit depth specifically, I could do that remotely.  

It would also be nice to be able to see on the app the current settings LosslessSwitcher detects and be able to double check that against what Apple Music says (without having to turn on a display and interact with the media player Mac Mini).

For my particular situation, an iOS app would be sufficient but if the companion app was Multiplatform and could run on a remote Mac computer as well that might also have value (more convenient than going into a Shared Screen remote control session).

**Comment by Robertsmania** [comment]
I've finally gotten around to working on this and have had pretty good initial success.  I've added a NetworkManager class and initialize it in AppDelegate.applicationDidFinishLaunching.  It provides a Bonjour listener and I can connect to it with a simple iOS app.  The iOS app shows the current sample rate and I've set it up to be able to toggle the defaults.userPreferIconStatusBarItem option as a proof of concept.  

My iOS companion app can get data from LosslessSwitcher on the Mac and send requests to change a parameter, good start!

I started my experiments with a vanilla clone of the main branch, and there everything has been working fine.

~~**I run into an issue importing the network framework on the improvements-1 branch.**~~

~~When I clone improvements-1 and attempt to import the network framework:~~
~~`import Network
`~~

~~It throws an error when I build:~~
~~`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/System/Library/Frameworks/Network.framework/Headers/Network.h:16:10 'Network/nw_object.h' file not found
`~~

~~I can get it to build and run if I adjust the LosslessSwitcher TARGETS Minimum Deployments and change it from the default of macOS 11.4 to 13.1 but that isn't really satisfactory.~~

~~In my studio, I'm using an old 2014 MacMini and it cant run macOS 13.1.  My laptop is newer and I can use it to develop on 13.2.1 with Xcode 14.2 (14C18).~~

~~I'm pretty new to developing on this platform and I've poked around the Xcode project settings but have not found anything that has let me build improvements-1 with the network framework and target any macOS lower than 13.1.~~

~~If I start a new macOS project, or modify the LosslessSwitch main branch code - then it works fine.  So it seems like something in the project configuration or package dependencies in improvements-1 is causing the issue?~~

I realized that there was a new Xcode, Version 14.3 (14E22b).

After updating, I find that I can now import the network framework just fine into the improvements-1 code.  

Sorry for the user error and any potential distraction.

I'll report back later when I've made more progress on the remote iOS companion app!

**Comment by Robertsmania** [comment]
It's not pretty, but it works!  Initial success.

![LLS_iOS_Remote_00](https://user-images.githubusercontent.com/11642124/231382523-c3b68f3d-95a3-4e99-9278-8aa09a7405e3.png)

**Comment by Robertsmania** [comment]
I have posted the initial version of my companion iOS app: LosslessSwitcherRemote:  
https://github.com/Robertsmania/LosslessSwitcherRemote

It does require a modified version of LosslessSwitcher with network support and some functional changes.  This version lets you toggle whether it auto switches, shows both the current and detected sample rates, and has an option to set the current sample rate to what is currently detected.  This is a fork from the main branch of LosslessSwitcher:
https://github.com/Robertsmania/LosslessSwitcherNetworkServer

Here's the updated menu for LosslessSwitcher running on the Mac.  The status bar values for C: and D: are the Current and Detected frequency rates.  In the menu you can manually apply the detected to set the current.  You can also toggle the Auto Switching option:

<img width="201" alt="LLS_MacOS_Menu_01" src="https://user-images.githubusercontent.com/11642124/231385169-37cf4e14-ae31-4cdb-87bb-5fb66aaa263b.png">

Here are some screenshots for the iOS companion app.

If you start the iOS app with no servers running this is the screen you get.  You can manually refresh and look for servers, but if none are running on your network this is as far as it goes:

![LLS_Remote_iOS_01](https://user-images.githubusercontent.com/11642124/231374790-244ad537-3f7a-4162-b49b-a50f89f9fcf8.png)

When you run the network enabled version of LosslessSwitcher, it appears on the network and the iOS app will detect it.  If it is the only service discovered, the iOS app will connect to it automatically. This screen shows the Manage Connection Details screen where you can see all the discovered services and pick the one you want to connect to:

![LLS_Remote_iOS_Connection_03](https://user-images.githubusercontent.com/11642124/231379797-41d052b9-9809-4385-bc72-2cab677830b0.png)

**This is the main UI when the app is in use:** 

![LLS_iOS_Remote_Main](https://user-images.githubusercontent.com/11642124/231378233-5f01ad1e-b5bb-44f0-85bb-9c73eefb0d96.png)

- At the top it shows the currently connected host name and the output device.  
- You can click the Show Connection Details button to get the screen above which shows all the discovered services and lets you connect to a different service.  
- The Refresh button updates all the data from the server, but is not normally needed as all the operations will update the UI with data that comes back from the server after any operation has been executed.
- The current and detected frequency rates are shown along with the current status of the auto-switching option.
- The buttons below that do pretty much what they say:
- Toggle Auto Switch turns on or off the auto switch feature.  When it is off, LosslessSwitcher will show the detected frequency rate, but the current setting will not change unless you use one of the manual options.
- Set Rate to Detected - mostly useful if auto switching is disabled.
- The picker shown has the supported frequency rates for the current output device.  You can select one and use the Set Rate to Selected button to apply it.

It supports multiple devices and multiple servers.  Each iOS device can select which server it wants to control and the servers do a best effort to send updates to all clients that are connected to them.

**Comment by Robertsmania** [comment]
Here is a video showing the LosslessSwitcherRemote iOS app in use:

[![Watch the video](https://img.youtube.com/vi/wjMEUtEg41s/0.jpg)](https://youtu.be/wjMEUtEg41s)

Also, the binary has been approved for distribution through TestFlight.  So you can get the source code and build it yourself:
https://github.com/Robertsmania/LosslessSwitcherRemote

Or, use this URL to join the beta test and get the iOS application installed on your device from TestFlight:
https://testflight.apple.com/join/QDtXe7tV

The app does require the forked/modified version of LosslessSwitcher with Network Server support:
https://github.com/Robertsmania/LosslessSwitcherNetworkServer

**Comment by Robertsmania** [comment]
@vincentneo I could really use your help.  I can no longer get LosslessSwitcher to build if "import Network" is included in the project.

When I started experimenting with adding the network support I saw the same issue in the improvements-1 branch where Xcode would not build and complained about 'Network/nw_object.h' file not found. At that time I was able to adjust it by changing the Target minimum macOS deployment.  Then discovered that there was a new version of Xcode, and it seemed to work fine with that and the original 11.4 target.  

I was able to work with that and get the network features working.

After some months I have come back to the project and find that with the current Xcode 14.3.1 (14E300c) I can no longer get it to build, no matter what I try.  I dont know if it's something that has changed in updated versions of the package dependencies, something in the current Xcode, macOS SDKs/frameworks, or just plain user error on my part - but it's really frustrating.

Repro Steps:

1. Start with a fresh clone of the current improvements-1 branch of LosslessSwitcher:
https://github.com/vincentneo/LosslessSwitcher.git

2. Build as it is and demonstrate that it compiles and runs.

3. Edit the AppDelegate.swift file to add "import Network" with the other import statements.

4. Try to build again and hopefully you see the same errors I do:
 ```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/Network.framework/Headers/Network.h:16:10 'Network/nw_object.h' file not found
 
Could not build Objective-C module 'Network'
```

Can you see what needs to be done to get the project to build with the network framework?

**Reply by vincentneo** [reply]
I tried building #90, which definitely imports Network.framework, and it seemingly worked fine for me. (no modifications other than change of development team for signing.)

That said, I am on Xcode 14.3 (14E222b), and in fact I did not realise there was a 14.3.1 until when you told me about the issue. Maybe that could be a factor? The use of private frameworks in the `improvements-1` branch could be a contributing factor towards such problems to be honest.

<img width="1201" alt="Screenshot 2023-08-06 at 3 40 48 PM" src="https://github.com/vincentneo/LosslessSwitcher/assets/23420208/46e925b7-fbd7-4ebe-bddb-4803210e044f">

**Reply by Robertsmania** [reply]
@vincentneo I never did get the project building with the Network framework imported after running into trouble in August, but would like to try again.

I'm currently on Sonoma 14.3.1 and Xcode 15.2.  

I cloned a clean version of the current improvements-1 branch, and that builds fine.  But if I add "imports Network" to QualityApp.swift then it fails to build, saying:
Did not find header 'nw_object.h' in framework 'Network' (loaded from '/System/Library/PrivateFrameworks')

<img width="1688" alt="Screenshot 2024-02-23 at 11 34 18 PM" src="https://github.com/vincentneo/LosslessSwitcher/assets/11642124/62650df0-1d1a-4a9f-a479-cd57806a0e1f">

Can you confirm if you get the same error or a successful build importing the Network framework?  If it is working for you, which versions of MacOS and Xcode are you on?  Thanks!

**Reply by vincentneo** [reply]
@Robertsmania 
<img width="1748" alt="Screenshot 2024-02-24 at 11 29 37 PM" src="https://github.com/vincentneo/LosslessSwitcher/assets/23420208/dea79e91-24b5-4c32-a611-d18a78d65499">
yeah looks like I'm seeing the same now (Xcode 15.2 / macOS Ventura 13.6.4). Will let you know if I get to fix it.

**Reply by Robertsmania** [reply]
I believe in you!

**Reply by vincentneo** [reply]
@Robertsmania go check out https://github.com/Robertsmania/LosslessSwitcherNetworkServer/pull/2 on your Mac, lets see if that fixes it

**Reply by Robertsmania** [reply]
Yes, thank you changing the Framework Search Path in Build Settings to look in PrivateFrameworks gets it to build. 

Its confusing because at the file system level /System/Library/PrivateFrameworks/Network.framework is a symbolic link that points to /System/Library/Network.framework - so I would have kind of expected it to resolve to the same actual files either way...
![image](https://github.com/vincentneo/LosslessSwitcher/assets/11642124/5f83ff49-6aea-44b9-a01f-a892ee49a4ab)

What I would also love to understand is why once its  built successfully with one project, any other that imported the Network.framework and was failing to build now does compile on the computer.  Its as though building once puts files outside the project that are then available so other projects that still have the search paths set with the PrivateFrameworks now find what they need?

**Reply by vincentneo** [reply]
My guess is maybe there could be some Xcode cache somewhere to speed things up but with a side effect of that odd behaviour?
Especially since system frameworks are unlikely to change.

**Reply by Robertsmania** [reply]
Seems like it can lead to frustrating "It works on my machine" situations between developers using the same code and tools but potentially different histories of what projects they've build/worked on previously.  But whatever, its working and I can move forward with what I was interested in playing with in the first place.

Thanks again for sharing the project!
